### ISSUE TITLE
[Feature Request]: Integrate Hammerspoon with fzf for fuzzy file finding

### ISSUE THREAD

**Issue by diego898** [Hammerspoon/hammerspoon#2370]
I'd love to use hammerspoon as an interface for https://github.com/junegunn/fzf to fuzzy find files on my file system.

**Comment by andweeb** [comment]
This was actually a project idea I've had in the back of my mind! Integrating fzf with Hammerspoon would be sick.

**Comment by myelsukov** [comment]
Check this: https://github.com/chipsenkbeil/choose

**Comment by diego898** [comment]
@myelsukov interesting! I think this workflow would more useful if we had fzf integrated into hammerspoon though

**Comment by latenitefilms** [comment]
I feel like this would probably be best as a Spoon, rather than as a bundled Hammerspoon extension?

**Comment by myelsukov** [comment]
I consider myself a power user of `fzf` (and `hs`), so I hope I understand what I am talking about :)

I seriously doubt that one can easily port advanced `fzf` features (preview, etc.) to the `hs`.
Also, `fzf` by its nature is a command line tool. It shines as a building block for making advanced shell scripts for the specific use cases (interactive command line history, interactive `grep`, interactive `cd`, interactive `ssh`, etc.) 

On the other hand, simple fuzzy search in the list is already implemented by the third party utilities such as `choose`.
I successfully use it with `hs`.
 
Of course the same functionality can be implemented natively in lua but why? To simplify installation process?

In any case I agree with @latenitefilms that implementation as a Spoon makes better sense.
It (Spoon) can be used to define proper interface and to prove usefulness of the new feature :).

Blind adding more and more features directly to `hs` would make it an incomprehensible unmaintainable monster. Look at Microsoft Office :)

**Comment by diego898** [comment]
@myelsukov agreed we dont want to fully port over advanced fzf features into hammerspoon. 

Can you describe more how you use choose with hs? that might be exactly what Im looking for

**Comment by myelsukov** [comment]
I am sorry I wasn't clear. I didn't mean to say that I use `fzf` with `hs`. I am using both, but separately :)
I use `choose` with `hs` for fuzzy searching.

Actually I have implemented a bash script I call **chooser** which uses `fzf`, `choose`, https://github.com/Cloudef/bemenu or pure bash depending on their availability and environment (X Windows, Mac OS, ssh in the terminal).

**Comment by diego898** [comment]
How do you `choose` with `hs`?

**Comment by myelsukov** [comment]
`choose` is a simple filter. It takes input from stdin, allows you to fuzzy find what you need and sends your choice to stdout.
Same as `fzf`
If I am not mistaken you can send data to the task stdin using `hs.task:setInput()` 
In the worst case you can write a shell script and call it from `hs`. I use it mostly as a shell script.

**Comment by andrewferrier** [comment]
@myelsukov just to be clear; `choose` is not exactly the same as `fzf`. `fzf` acts as a pure Unix filter, which is to say it acts asynchronously (you can continue feeding stdin whilst the user is type). `choose` is blocking; it waits for stdout to finish before even presenting the options. (I've already opened an issue on choose re: this: https://github.com/chipsenkbeil/choose/issues/18). This makes it much less attractive for some choices (e.g. where there is a list that takes some time to generate, but often the items the user wants are early on). I'm not bashing on `choose`; currently it's the only GUI-ready tool similar to `fzf` for MacOS, and the author's done a great job, but it does have that key limitation. Ideally I would rather see a similar capability built-in to Hammerspoon; but appreciate that's a lot of work, and I'm not volunteering :)

**Comment by asmagill** [comment]
I may be missing something obvious here, but how exactly do you see Hammerspoon and fzf interacting? What would be the work flow?

If it's a matter of putting up a list of options for the user to select from, there's `hs.chooser` (in fact, I think it was originally inspired by `choose`, or something similar) and while it doesn't currently do fuzzy matching (though it is case insensitive), there are enough callback hooks that it probably could be done. I personally want to add regular expression matching support to `hs.chooser` at some point.

Are you thinking of a function which would take a list of items and a search term and return a list of ranked matches using a fuzzy search algorithm? IMHO, that would be a useful addition to Hammerspoon at some point and could probably integrate with `hs.chooser` quite nicely.

Are you thinking of one of the command line versions piping its data into Hammerspoon for some sort of manipulation, then returning it for display within the terminal? That's outside the scope of Hammerspoon and probably slow, even if you could do it with judicious use of the `hs` command line tool.

Or, as I alluded to at the top, am I missing the intended use case entirely?

**Comment by diego898** [comment]
At the highest level - I use two tools on my mac - Alfred to navigate to files/folders (but it doesn't support fuzzy finding), and hammerspoon for various other tasks. One particularly useful tool I use in hs, is [Anycomplete](https://github.com/nathancahill/Anycomplete) . 

I was just wondering if I could also use this workflow - typing a key combo to brining up a bar, I type, and its populated with fuzzy found files, could be incorporated directly into hammerspoon, allowing me to get rid of Alfred.

**Comment by myelsukov** [comment]
@andrewferrier 
> fzf acts as a pure Unix filter, which is to say it acts asynchronously

To my knowledge, Unix filters never reprocess already processed stream data. In this sense `fzf` is not a pure Unix filter :)
Unix filters don't _have_ to be asynchronous. For example, `sort` filter reads the whole input before producing any output.

**Comment by myelsukov** [comment]
@diego898 You can do what you want using shell script. Just assign a hotkey in `hs` to launch that script.
I use fuzzy search to find an entry in my [pass](https://www.passwordstore.org/) database and it works exactly how you describe it. I press a hotkey, then choose the entry in the pop-up window. Check this: https://gist.github.com/myelsukov/df4e99c89036ee66033bfb491a010676

**Comment by NightMachinary** [comment]
[**@asmagill**](https://github.com/asmagill) commented on [Jul 13, 2020, 11:02 PM GMT+4:30](https://github.com/Hammerspoon/hammerspoon/issues/2370#issuecomment-657721594 "2020-07-13T18:32:40Z - Replied by Github Reply Comments"):
> I may be missing something obvious here, but how exactly do you see Hammerspoon and fzf interacting? What would be the work flow?
> 
> If it's a matter of putting up a list of options for the user to select from, there's `hs.chooser` (in fact, I think it was originally inspired by `choose`, or something similar) and while it doesn't currently do fuzzy matching (though it is case insensitive), there are enough callback hooks that it probably could be done. I personally want to add regular expression matching support to `hs.chooser` at some point.
> 
> Are you thinking of a function which would take a list of items and a search term and return a list of ranked matches using a fuzzy search algorithm? IMHO, that would be a useful addition to Hammerspoon at some point and could probably integrate with `hs.chooser` quite nicely.
> 
> Are you thinking of one of the command line versions piping its data into Hammerspoon for some sort of manipulation, then returning it for display within the terminal? That's outside the scope of Hammerspoon and probably slow, even if you could do it with judicious use of the `hs` command line tool.
> 
> Or, as I alluded to at the top, am I missing the intended use case entirely?

I am thinking of having workflows like this:
1. Register key in hammerspoon
2. Feed some data to the fuzzy chooser (e.g., Chrome's history)
3. Choose the desired item
4. Profit

Basically, the usecase should be exactly the same as `hs.chooser`, but with fzf as a search backend. I am taking a look at the hooks `hs.chooser` provides to create this, and I'll share my results here.

**Comment by NightMachinary** [comment]
I [implemented](https://github.com/NightMachinary/.shells/blob/162f87ca8a368290a02f738c663bfabf2c1ff18e/.hammerspoon/init.lua#L279) a Chrome history browser that uses fzf for the filtering part:

![image](https://user-images.githubusercontent.com/36224762/96940158-d3477100-14db-11eb-8a55-021fa50a172d.png)

I use these functions to connect to my shell wrappers:

```
function exec_raw(cmd)
  local f = assert(io.popen(cmd, 'r'))
  local s = assert(f:read('*a'))
  f:close()
  return (s)
end
function exec(cmd)
  -- hs.alert.show(cmd, 20)
  return trim1(exec_raw(cmd))
end
function trim1(s)
  return (s:gsub("^%s*(.-)%s*$", "%1"))
end
function brishzeval2(cmd)
  local cmdq = ("brishz_quote=y /usr/local/bin/brishz.dash %q"):format(cmd)
  return exec(cmdq)
end
```

`/usr/local/bin/brishz.dash` can be found at https://github.com/NightMachinary/.shells/blob/master/scripts/zsh/wrappers/brishz/brishz.dash. It is itself a wrapper around my zsh HTTP API, https://github.com/NightMachinary/BrishGarden. (The reason it doesn't eval the code directly is that I have a lot of zsh code with complex interdependencies, which take some seconds to load. So I prefer to use a server that load-balances my zsh eval requests between some ready workers.) The functions ultimately called (the real backend, which uses fzf under the hood) reside at https://github.com/NightMachinary/.shells/blob/master/scripts/zsh/auto-load/others/web.zsh, https://github.com/NightMachinary/.shells/blob/master/scripts/zsh/basic/auto-load/fuzzy.zsh.

One problem I could not solve was making the requests to the backend async, so that hammerspoon does not wait for them. I ultimately implemented a timer that updates the entries with a delay, so as to allow me to type quickly and not wait for updates on each char typed. This is not so bad except things aren't as snappy as I like them. (fzf itself is very fast.)

**Comment by NightMachinary** [comment]
BTW, why aren't the entries filling the whole width?

> ![image](https://user-images.githubusercontent.com/36224762/96940158-d3477100-14db-11eb-8a55-021fa50a172d.png)
